## Import Libraries

In [2]:
import pandas as pd
import numpy as np
import os

## Import Data

In [4]:
#path to main folder
path = r'C:\Users\steve\Documents\Olist Marketplace Analysis'

# import order items file
items_df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'olist_order_items_dataset.csv'), index_col = False)

## Analysis

#### 01. Content + Shape

In [7]:
items_df.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [8]:
items_df.tail()

order_id  order_item_id  \
112645  fffc94f6ce00a00581880bf54a75a037              1   
112646  fffcd46ef2263f404302a634eb57f7eb              1   
112647  fffce4705a9662cd70adb13d4a31832d              1   
112648  fffe18544ffabc95dfada21779c9644f              1   
112649  fffe41c64501cc87c801fd61db3f6244              1   

                              product_id                         seller_id  \
112645  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
112646  32e07fd915822b0765e448c4dd74c828  f3c38ab652836d21de61fb8314b69182   
112647  72a30483855e2eafc67aee5dc2560482  c3cfdc648177fdbbbb35635a37472c53   
112648  9c422a519119dcad7575db5af1ba540e  2b3e4a2a3ea8e01938cabda2a3e5cc79   
112649  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

        shipping_limit_date   price  freight_value  
112645  2018-05-02 04:11:01  299.99          43.41  
112646  2018-07-20 04:31:48  350.00          36.53  
112647  2017-10-30 17:14:25   99.90          16.95  
112648  2017-08-21 00:04:32   55.99           8.72  
112649  2018-06-12 17:10:13   43.00          12.79

In [9]:
# reorder so that order_item_id is after product_id (to increase legibility)
new_column_order = ['order_id', 'seller_id', 'product_id', 'shipping_limit_date', 'order_item_id', 'price', 'freight_value']
items_df = items_df[new_column_order]

items_df.head(1)

order_id                         seller_id  \
0  00010242fe8c5a6d1ba2dd792cb16214  48436dade18ac8b2bce089ec2a041202   

                         product_id  shipping_limit_date  order_item_id  \
0  4244733e06e7ecb4970a6e2683c13e61  2017-09-19 09:45:35              1   

   price  freight_value  
0   58.9          13.29

In [10]:
items_df.shape

(112650, 7)

In [11]:
items_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   seller_id            112650 non-null  object 
 2   product_id           112650 non-null  object 
 3   shipping_limit_date  112650 non-null  object 
 4   order_item_id        112650 non-null  int64  
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [12]:
items_df.describe()

order_item_id          price  freight_value
count  112650.000000  112650.000000  112650.000000
mean        1.197834     120.653739      19.990320
std         0.705124     183.633928      15.806405
min         1.000000       0.850000       0.000000
25%         1.000000      39.900000      13.080000
50%         1.000000      74.990000      16.260000
75%         1.000000     134.900000      21.150000
max        21.000000    6735.000000     409.680000

#### 02. Derive Columns

In [14]:
# CREATE COL that shows the total price of items in an order

# group by 'order_id' and sum the 'price' for each order
order_total_price = items_df.groupby('order_id')['price'].sum().reset_index()
# rename the column to 'total_price'
order_total_price = order_total_price.rename(columns={'price': 'total_price'})
# merge the col back into the df
items_df = pd.merge(items_df, order_total_price, on='order_id', how='left')

items_df.head(3)

order_id                         seller_id  \
0  00010242fe8c5a6d1ba2dd792cb16214  48436dade18ac8b2bce089ec2a041202   
1  00018f77f2f0320c557190d7a144bdd3  dd7ddc04e1b6c2c614352b383efe2d36   
2  000229ec398224ef6ca0657da4fc703e  5b51032eddd242adc84c38acab88f23d   

                         product_id  shipping_limit_date  order_item_id  \
0  4244733e06e7ecb4970a6e2683c13e61  2017-09-19 09:45:35              1   
1  e5f2d52b802189ee658865ca93d83a8f  2017-05-03 11:05:13              1   
2  c777355d18b72b67abbeef9df44fd0fd  2018-01-18 14:48:30              1   

   price  freight_value  total_price  
0   58.9          13.29         58.9  
1  239.9          19.93        239.9  
2  199.0          17.87        199.0

In [15]:
# create col that shows the total frieght value per order
order_total_freight = items_df.groupby('order_id')['freight_value'].sum().reset_index() # group by order id & sum freight value
order_total_freight = order_total_freight.rename(columns={'freight_value': 'total_freight'}) # rename col
items_df = pd.merge(items_df, order_total_freight, on='order_id', how='left') # merge col into items_df

items_df.head(3)

order_id                         seller_id  \
0  00010242fe8c5a6d1ba2dd792cb16214  48436dade18ac8b2bce089ec2a041202   
1  00018f77f2f0320c557190d7a144bdd3  dd7ddc04e1b6c2c614352b383efe2d36   
2  000229ec398224ef6ca0657da4fc703e  5b51032eddd242adc84c38acab88f23d   

                         product_id  shipping_limit_date  order_item_id  \
0  4244733e06e7ecb4970a6e2683c13e61  2017-09-19 09:45:35              1   
1  e5f2d52b802189ee658865ca93d83a8f  2017-05-03 11:05:13              1   
2  c777355d18b72b67abbeef9df44fd0fd  2018-01-18 14:48:30              1   

   price  freight_value  total_price  total_freight  
0   58.9          13.29         58.9          13.29  
1  239.9          19.93        239.9          19.93  
2  199.0          17.87        199.0          17.87

In [16]:
# create col that shows the order total (the total price + total shipping)
items_df['order_total'] = items_df['total_price'] + items_df['total_freight']
items_df.head(3)

order_id                         seller_id  \
0  00010242fe8c5a6d1ba2dd792cb16214  48436dade18ac8b2bce089ec2a041202   
1  00018f77f2f0320c557190d7a144bdd3  dd7ddc04e1b6c2c614352b383efe2d36   
2  000229ec398224ef6ca0657da4fc703e  5b51032eddd242adc84c38acab88f23d   

                         product_id  shipping_limit_date  order_item_id  \
0  4244733e06e7ecb4970a6e2683c13e61  2017-09-19 09:45:35              1   
1  e5f2d52b802189ee658865ca93d83a8f  2017-05-03 11:05:13              1   
2  c777355d18b72b67abbeef9df44fd0fd  2018-01-18 14:48:30              1   

   price  freight_value  total_price  total_freight  order_total  
0   58.9          13.29         58.9          13.29        72.19  
1  239.9          19.93        239.9          19.93       259.83  
2  199.0          17.87        199.0          17.87       216.87

In [17]:
# run stats again for new cols
items_df.describe()

order_item_id          price  freight_value    total_price  \
count  112650.000000  112650.000000  112650.000000  112650.000000   
mean        1.197834     120.653739      19.990320     152.959595   
std         0.705124     183.633928      15.806405     257.339594   
min         1.000000       0.850000       0.000000       0.850000   
25%         1.000000      39.900000      13.080000      49.000000   
50%         1.000000      74.990000      16.260000      91.550000   
75%         1.000000     134.900000      21.150000     164.990000   
max        21.000000    6735.000000     409.680000   13440.000000   

       total_freight    order_total  
count  112650.000000  112650.000000  
mean       27.285033     180.244628  
std        33.218827     272.829911  
min         0.000000       9.590000  
25%        14.292500      65.620000  
50%        18.160000     114.440000  
75%        29.220000     195.337500  
max      1794.960000   13664.080000

#### 03. Value Counts

Order item id

In [20]:
# get counts of order items
items_df['order_item_id'].value_counts(dropna = False)

order_item_id
1     98666
2      9803
3      2287
4       965
5       460
6       256
7        58
8        36
9        28
10       25
11       17
12       13
13        8
14        7
15        5
16        3
17        3
18        3
19        3
20        3
21        1
Name: count, dtype: int64

**Unique Order Item IDs** = 21

Product id

In [23]:
# get counts of cities
items_df['product_id'].value_counts(dropna = False)

product_id
aca2eb7d00ea1a7b8ebd4e68314663af    527
99a4788cb24856965c36a24e339b6058    488
422879e10f46682990de24d770e7f83d    484
389d119b48cf3043d311335e499d9c6b    392
368c6c730842d78016ad823897a372db    388
                                   ... 
dbc42f8ad8afd0147a5a46be630781f8      1
5d86ed92606c80a89e547cdba3946812      1
0b8750fa58fefe42cf03a9e2e4b8fce6      1
a4758e74e2cdb74d0c4db83594ca14a3      1
350688d9dc1e75ff97be326363655e01      1
Name: count, Length: 32951, dtype: int64

**Unique Product IDs** = 32,951

Seller id

In [26]:
# get counts of states
items_df['seller_id'].value_counts(dropna = False)

seller_id
6560211a19b47992c3666cc44a7e94c0    2033
4a3ca9315b744ce9f8e9374361493884    1987
1f50f920176fa81dab994f9023523100    1931
cc419e0650a3c5ba77189a1882b7556a    1775
da8622b14eb17ae2831f4ac5b9dab84a    1551
                                    ... 
400f221ab83037c0ea78f1267da9624d       1
1746c1ae87ac63d530c7c943d0ac42e2       1
16bdc8cefd0e32a6f0824d296c5ad14a       1
ace601496e211a664100321f038f50af       1
13d95f0f6f73943d4ceffad0fc2cd32c       1
Name: count, Length: 3095, dtype: int64

**Unique Seller IDs** = 3,095

### CONSISTENCY CHECKS

#### 01. Mixed-Type Data

In [30]:
# check if there are any mixed-type columns
for col in items_df.columns.tolist():
    weird = (items_df[[col]].map(type) != items_df[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (items_df[weird]) > 0:
        print(col)
    else:
        print('not mixed col')

not mixed col
not mixed col
not mixed col
not mixed col
not mixed col
not mixed col
not mixed col
not mixed col
not mixed col
not mixed col


**No mixed columns**

#### 02. Missing Values

In [33]:
# returns number of missing data by column
items_df.isnull().sum()

order_id               0
seller_id              0
product_id             0
shipping_limit_date    0
order_item_id          0
price                  0
freight_value          0
total_price            0
total_freight          0
order_total            0
dtype: int64

**No missing values**

#### 03. Duplicates

In [36]:
# creates a subet containing only the full duplicates
df_dups = items_df[items_df.duplicated()]
df_dups

Empty DataFrame
Columns: [order_id, seller_id, product_id, shipping_limit_date, order_item_id, price, freight_value, total_price, total_freight, order_total]
Index: []

**No full duplicates**

### Export Data

In [39]:
items_df.to_csv(os.path.join(path, '02 Data', 'Prepared Data','order_items_checked.csv'))